File: tests/oracle/options_wiring_1204.cc
/*
 * OPTIONS-SPLITDATATYPE-WIRING-0002: locked Ghidra 12.0.4 behavioral oracle
 * for ActionDatabase::toggleAction (action.cc:1036-1053) driven through the
 * real OptionSplitDatatypes::apply forwarding (options.cc:1007-1016).
 *
 * The fixture builds a synthetic universal root (head leaves in "base", a
 * nested "body" group holding the "splitcopy"/"splitpointer"/"merge" leaves),
 * registers it through the production ActionDatabase, derives the "decompile"
 * root via setCurrent, then runs the production OptionSplitDatatypes option
 * over parameter triples covering: full-off, struct-only (splitpointer
 * removed), struct+pointer, re-assertion of an identical configuration
 * (toggleAction still replaces the root object), a p1 LowlevelError (no
 * mutation at all), a p2 LowlevelError (partial configuration assignment,
 * toggle block never reached), and an array-only re-enable.  Production
 * bodies are linked unchanged from the locked libdecomp archive; the
 * private/protected visibility macros only enable state observation, and a
 * minimal Architecture subclass stubs the pure virtuals that the option
 * path never calls.
 */

#include <bits/stdc++.h>
#define class struct
#define private public
#define protected public
#include "architecture.hh"
#include "options.hh"
#undef class
#undef private
#undef protected

using namespace ghidra;
using std::cout;
using std::runtime_error;

// The inherited flags word carries this leaf's construction ordinal (a
// monotonically increasing counter) solely as an allocator-independent
// identity channel: every clone is freshly constructed, so the ordinal
// changes exactly when the owning root was re-derived/replaced.
static int next_script_ordinal = 100;

class ScriptAction final : public Action {
public:
  ScriptAction(const string &g, const string &nm)
      : Action((uint4)next_script_ordinal++, nm, g) {}

  Action *clone(const ActionGroupList &grouplist) const override {
    if (!grouplist.contains(getGroup()))
      return (Action *)0;
    return new ScriptAction(getGroup(), getName());
  }

  int4 apply(Funcdata &) override { return 0; }
};

class TestArch : public Architecture {
public:
  TestArch(void) : Architecture() {}
  virtual ~TestArch(void) {}

  void printMessage(const string &message) const override {
    (void)message;
  }
  Translate *buildTranslator(DocumentStorage &) override { return (Translate *)0; }
  void buildLoader(DocumentStorage &) override {}
  PcodeInjectLibrary *buildPcodeInjectLibrary(void) override {
    return (PcodeInjectLibrary *)0;
  }
  void buildTypegrp(DocumentStorage &) override {}
  void buildCoreTypes(DocumentStorage &) override {}
  void buildCommentDB(DocumentStorage &) override {}
  void buildStringManager(DocumentStorage &) override {}
  void buildConstantPool(DocumentStorage &) override {}
  void buildContext(DocumentStorage &) override {}
  void buildSymbols(DocumentStorage &) override {}
  void buildSpecFile(DocumentStorage &) override {}
  void modifySpaces(Translate *) override {}
  void resolveArchitecture(void) override {}
};

static string render_group_children(const ActionGroup *grp) {
  string out;
  for (size_t i = 0; i < grp->list.size(); ++i) {
    if (i != 0)
      out += ',';
    const ActionGroup *sub = dynamic_cast<const ActionGroup *>(grp->list[i]);
    if (sub != (const ActionGroup *)0) {
      out += sub->getName();
      out += '[';
      out += render_group_children(sub);
      out += ']';
    } else {
      out += grp->list[i]->getName();
      out += '@';
      out += grp->list[i]->getGroup();
    }
  }
  return out;
}

static void emit_state(TestArch &arch, const string &tag) {
  const ActionDatabase &db = arch.allacts;
  const ActionGroupList &grp = db.getGroup("decompile");
  string members;
  for (set<string>::const_iterator it = grp.list.begin();
       it != grp.list.end(); ++it) {
    if (it != grp.list.begin())
      members += ',';
    members += *it;
  }
  const ActionGroup *root =
      dynamic_cast<const ActionGroup *>(db.getCurrent());
  if (root == (const ActionGroup *)0)
    throw runtime_error("current root is not an ActionGroup");
  cout << "state=" << tag << " config=" << arch.split_datatype_config
       << " current=" << db.getCurrentName() << " mapsize="
       << db.actionmap.size() << " members=" << members
       << " tree=" << render_group_children(root) << '\n';
}

static int first_leaf_ordinal(const Action *root) {
  const ActionGroup *grp = dynamic_cast<const ActionGroup *>(root);
  if (grp == (const ActionGroup *)0 || grp->list.empty())
    throw runtime_error("root is not a non-empty ActionGroup");
  return (int)grp->list[0]->flags;
}

int main(void) {
  // ghidra_process.cc:523 registers the linked capabilities (the C print
  // language) before any Architecture construction.
  CapabilityPoint::initializeAll();

  TestArch arch;

  // Synthetic universal root: head leaves + nested body group holding the
  // split-relevant leaves at their registration slots.
  ActionRestartGroup *universal =
      new ActionRestartGroup(Action::rule_onceperfunc, "universal", 1);
  universal->addAction(new ScriptAction("base", "start"));
  universal->addAction(new ScriptAction("base", "stop"));
  {
    ActionGroup *body = new ActionGroup(0, "body");
    body->addAction(new ScriptAction("splitcopy", "splitcopy"));
    body->addAction(new ScriptAction("splitpointer", "splitpointer"));
    body->addAction(new ScriptAction("merge", "merge"));
    universal->addAction(body);
  }
  arch.allacts.registerAction("universal", universal);

  const char *decompile_members[] = {"base", "splitcopy", "splitpointer",
                                     "merge", (const char *)0};
  arch.allacts.setGroup("decompile", decompile_members);
  arch.allacts.setCurrent("decompile");
  emit_state(arch, "init");

  struct Case {
    const char *p1;
    const char *p2;
    const char *p3;
  };
  const Case cases[] = {
      {"", "", ""},
      {"struct", "", ""},
      {"struct", "", "pointer"},
      {"array", "struct", "pointer"},
      {"array", "struct", "pointer"},
      {"bogus", "", ""},
      {"struct", "bogus", ""},
      {"", "array", ""},
  };

  OptionSplitDatatypes option;
  for (size_t i = 0; i < sizeof(cases) / sizeof(cases[0]); ++i) {
    const int universal_before =
        first_leaf_ordinal(arch.allacts.getAction("universal"));
    const int root_before = first_leaf_ordinal(arch.allacts.getCurrent());

    bool threw = false;
    string result;
    try {
      result = option.apply(&arch, cases[i].p1, cases[i].p2, cases[i].p3);
    } catch (const LowlevelError &e) {
      threw = true;
      result = e.explain;
    }

    cout << "apply=" << i << " args=" << cases[i].p1 << '|' << cases[i].p2
         << '|' << cases[i].p3 << " threw=" << (threw ? 1 : 0)
         << " result=" << result << " config=" << arch.split_datatype_config
         << '\n';
    emit_state(arch, to_string(i));

    const int universal_after =
        first_leaf_ordinal(arch.allacts.getAction("universal"));
    const int root_after = first_leaf_ordinal(arch.allacts.getCurrent());
    cout << "identity=" << i << " universal_changed="
         << (universal_before != universal_after ? 1 : 0) << " root_changed="
         << (root_before != root_after ? 1 : 0)
         << " universal_firstid=" << universal_after
         << " root_firstid=" << root_after << '\n';
  }
  return 0;
}

File: tests/oracle/options_wiring_1204.rs
//! OPTIONS-SPLITDATATYPE-WIRING-0002 Rust comparand (re-pinned by
//! OPTIONS-SPLITDATATYPE-WIRING-0003).
//!
//! Calls the production `OptionSplitDatatypes::apply`
//! (`rugra::options::OptionSplitDatatypes`) directly: the configuration
//! bits, partial-assignment error ordering, return string, and the
//! internal `allacts` forwarding (options.cc:1007-1016) — the two
//! `ActionDatabase::toggle_action` calls on the current root — all run
//! inside the production body. No hand-written expected output is
//! embedded.

use rugra::action::{Action, ActionDatabase, ActionGroup, ActionGroupList, ActionRestartGroup};
use rugra::arch::Architecture;
use rugra::options::ArchOption;

// The inherited flags word carries this leaf's construction ordinal (a
// monotonically increasing counter) solely as an allocator-independent
// identity channel: every clone is freshly constructed, so the ordinal
// changes exactly when the owning root was re-derived/replaced.
static NEXT_SCRIPT_ORDINAL: std::sync::atomic::AtomicU32 =
    std::sync::atomic::AtomicU32::new(100);

struct ScriptAction {
    group: &'static str,
    name: &'static str,
    ordinal: u32,
}

impl ScriptAction {
    fn new(group: &'static str, name: &'static str) -> Self {
        Self {
            group,
            name,
            ordinal: NEXT_SCRIPT_ORDINAL
                .fetch_add(1, std::sync::atomic::Ordering::SeqCst),
        }
    }
}

impl Action for ScriptAction {
    fn apply(&mut self, _fd: &mut rugra::funcdata::Funcdata) -> rugra::Result<i32> {
        Ok(0)
    }
    fn get_name(&self) -> &str {
        self.name
    }
    fn get_flags(&self) -> u32 {
        self.ordinal
    }
    fn clone_for_groups(&self, grouplist: &ActionGroupList) -> Option<Box<dyn Action>> {
        if !grouplist.contains(self.group) {
            return None;
        }
        Some(Box::new(Self::new(self.group, self.name)))
    }
}

fn render_group_children(group: &ActionGroup) -> String {
    group
        .child_actions()
        .iter()
        .enumerate()
        .map(|(index, child)| match child.as_action_group() {
            Some(sub) => format!("{}[{}]", sub.get_name_str(), render_group_children(sub)),
            None => format!("{}@{}", child.get_name(), group.child_group(index)),
        })
        .collect::<Vec<_>>()
        .join(",")
}

fn emit_state(arch: &Architecture, tag: &str) {
    let db = arch.allacts.as_ref().expect("allacts populated").read().expect("allacts read lock");
    let members = db
        .get_group("decompile")
        .expect("decompile grouplist exists")
        .member_names()
        .join(",");
    let tree = render_group_children(db.get_current().as_action_group().expect("current root is a group"));
    println!(
        "state={tag} config={} current={} mapsize={} members={members} tree={tree}",
        arch.split_datatype_config,
        db.get_current_name(),
        db.actionmap_size(),
    );
}

fn apply_case(arch: &mut Architecture, p1: &str, p2: &str, p3: &str) -> (bool, String) {
    // Direct production apply (options.cc:999-1020): configuration bits,
    // partial-assignment error ordering, return string, and the internal
    // allacts forwarding (options.cc:1007-1016) that
    // OPTIONS-SPLITDATATYPE-WIRING-0003 wired into
    // `OptionSplitDatatypes::apply`. Since that wiring, the comparand is a
    // single production call, mirroring the oracle fixture's
    // `option.apply(&arch, ...)` entrypoint.
    let result = rugra::options::OptionSplitDatatypes.apply(arch, p1, p2, p3);
    match result.strip_prefix("LowlevelError: ") {
        Some(message) => (true, message.to_string()),
        None => (false, result),
    }
}

fn first_leaf_ordinal(action: Option<&dyn Action>) -> u32 {
    action
        .and_then(Action::as_action_group)
        .and_then(|group| group.child_actions().first())
        .map(|leaf| leaf.get_flags())
        .expect("root is a non-empty ActionGroup")
}

fn main() {
    let mut arch = Architecture::new();

    // Synthetic universal root: head leaves + nested body group holding the
    // split-relevant leaves at their registration slots (mirrors the oracle
    // fixture's universal construction).
    let mut universal = ActionRestartGroup::new("universal", rugra::action::action_flags::RULE_ONCEPERFUNC, 1);
    universal.add_action_in_group(Box::new(ScriptAction::new("base", "start")), "base");
    universal.add_action_in_group(Box::new(ScriptAction::new("base", "stop")), "base");
    let mut body = ActionGroup::new("body");
    body.add_action_in_group(Box::new(ScriptAction::new("splitcopy", "splitcopy")), "splitcopy");
    body.add_action_in_group(Box::new(ScriptAction::new("splitpointer", "splitpointer")), "splitpointer");
    body.add_action_in_group(Box::new(ScriptAction::new("merge", "merge")), "merge");
    universal.add_action(Box::new(body));

    let mut db = ActionDatabase::new();
    db.register_action(Box::new(universal));
    db.set_group("decompile", &["base", "splitcopy", "splitpointer", "merge"]);
    db.set_current("decompile");
    arch.allacts = Some(std::sync::Arc::new(std::sync::RwLock::new(db)));
    emit_state(&arch, "init");

    let cases: [(&str, &str, &str); 8] = [
        ("", "", ""),
        ("struct", "", ""),
        ("struct", "", "pointer"),
        ("array", "struct", "pointer"),
        ("array", "struct", "pointer"),
        ("bogus", "", ""),
        ("struct", "bogus", ""),
        ("", "array", ""),
    ];

    for (i, (p1, p2, p3)) in cases.iter().enumerate() {
        let db = arch.allacts.as_ref().expect("allacts populated").read().expect("allacts read lock");
        let universal_before = first_leaf_ordinal(db.get_action("universal"));
        let root_before = first_leaf_ordinal(Some(db.get_current()));
        drop(db);

        let (threw, result) = apply_case(&mut arch, p1, p2, p3);
        println!("apply={i} args={p1}|{p2}|{p3} threw={} result={result} config={}", u8::from(threw), arch.split_datatype_config);
        emit_state(&arch, &i.to_string());

        let db = arch.allacts.as_ref().expect("allacts populated").read().expect("allacts read lock");
        let universal_after = first_leaf_ordinal(db.get_action("universal"));
        let root_after = first_leaf_ordinal(Some(db.get_current()));
        println!(
            "identity={i} universal_changed={} root_changed={} universal_firstid={universal_after} root_firstid={root_after}",
            u8::from(universal_before != universal_after),
            u8::from(root_before != root_after),
        );
    }
}
